engine/access/rest/websockets/controller.go

```diff
@@ -0,0 +1,200 @@
+			if err != nil {
+				c.logger.Debug().Err(err).Msg("error parsing and validating client message")
+				return
+			}
+
+			if err := c.handleAction(ctx, baseMsg.Action, validatedMsg); err != nil {
+				c.logger.Warn().Err(err).Str("action", baseMsg.Action).Msg("error handling action")
+			}
+		}
+	}
+}
+
+func (c *Controller) readMessage() (json.RawMessage, error) {
+	var message json.RawMessage
+	if err := c.conn.ReadJSON(&message); err != nil {
+		return nil, fmt.Errorf("error reading JSON from client: %w", err)
+	}
+	return message, nil
+}
+
+func (c *Controller) parseAndValidateMessage(message json.RawMessage) (models.BaseMessageRequest, interface{}, error) {
+	var baseMsg models.BaseMessageRequest
+	if err := json.Unmarshal(message, &baseMsg); err != nil {
+		return models.BaseMessageRequest{}, nil, fmt.Errorf("error unmarshalling base message: %w", err)
+	}
+
+	var validatedMsg interface{}
+	switch baseMsg.Action {
+	case "subscribe":
+		var subscribeMsg models.SubscribeMessageRequest
+		if err := json.Unmarshal(message, &subscribeMsg); err != nil {
+			return baseMsg, nil, fmt.Errorf("error unmarshalling subscribe message: %w", err)
+		}
+		//TODO: add validation logic for `topic` field
+		validatedMsg = subscribeMsg
+
+	case "unsubscribe":
+		var unsubscribeMsg models.UnsubscribeMessageRequest
+		if err := json.Unmarshal(message, &unsubscribeMsg); err != nil {
+			return baseMsg, nil, fmt.Errorf("error unmarshalling unsubscribe message: %w", err)
+		}
+		validatedMsg = unsubscribeMsg
+
+	case "list_subscriptions":
+		var listMsg models.ListSubscriptionsMessageRequest
+		if err := json.Unmarshal(message, &listMsg); err != nil {
+			return baseMsg, nil, fmt.Errorf("error unmarshalling list subscriptions message: %w", err)
+		}
+		validatedMsg = listMsg
+
+	default:
+		c.logger.Debug().Str("action", baseMsg.Action).Msg("unknown action type")
+		return baseMsg, nil, fmt.Errorf("unknown action type: %s", baseMsg.Action)
+	}
+
+	return baseMsg, validatedMsg, nil
+}
+
+func (c *Controller) handleAction(ctx context.Context, action string, message interface{}) error {
+	switch action {
+	case "subscribe":
+		c.handleSubscribe(ctx, message.(models.SubscribeMessageRequest))
+	case "unsubscribe":
+		c.handleUnsubscribe(ctx, message.(models.UnsubscribeMessageRequest))
+	case "list_subscriptions":
+		c.handleListSubscriptions(ctx, message.(models.ListSubscriptionsMessageRequest))
+	default:
+		return fmt.Errorf("unknown action type: %s", action)
+	}
+	return nil
+}
+
+func (c *Controller) handleSubscribe(ctx context.Context, msg models.SubscribeMessageRequest) {
+	dp := c.dataProvidersFactory.NewDataProvider(ctx, c.communicationChannel, msg.Topic)
+	c.dataProviders.Add(dp.ID(), dp)
+	dp.Run(ctx)
+}
+
+func (c *Controller) handleUnsubscribe(ctx context.Context, msg models.UnsubscribeMessageRequest) {
+	id, err := uuid.Parse(msg.ID)
+	if err != nil {
+		c.logger.Debug().Err(err).Msg("error parsing message ID")
```

Comment:
I added an appropriate TODO for every handler. I wanted to do it in a separate task. 
We also have 2 more or less related issues for this https://github.com/onflow/flow-go/issues/6724 https://github.com/onflow/flow-go/issues/6642

I want to add tests alongside each request/response handler but it will require a lot of changes. This PR is already big enough.